File: AtCoder/ABC/082/cpp/D_FT_Robot_rec.cpp
// Answers are correct but TLE (Time Limit Exceeded)

#include <iostream>
#include <stdio.h>
#include <string>

using namespace std;

int M[4][2] = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

bool f(string s, int i, int N, int x, int y, int _x, int _y, int direction) {
  while (i < N && s[i] == 'F') {
    _x += M[direction][0];
    _y += M[direction][1];
    ++i;
  }
  if (i >= N) {
    return _x == x && _y == y;
  } else {
    bool res0 = f(s, i + 1, N, x, y, _x, _y, (direction + 1) % 4);
    bool res1 = f(s, i + 1, N, x, y, _x, _y, (direction + 3) % 4);
    return res0 || res1;
  }
}

int main() {
  string s;
  cin >> s;
  int x, y;
  cin >> x >> y;

  if (f(s, 0, s.size(), x, y, 0, 0, 0)) {
    printf("Yes\n");
  } else {
    printf("No\n");
  }
}

File: AtCoder/ABC/082/python/D_FT_Robot_rec.py
# LTE

S_MAX = 8_000

M = [[1, 0], [0, 1], [-1, 0], [0, -1]]

s = input()
x, y = [int(i) for i in input().split()]
N = len(s)


def f(s, i, d, _x, _y):
    while i < N and s[i] == 'F':
        _x += M[d][0]
        _y += M[d][1]
        i += 1
    if i >= N:
        return _x == x and _y == y
    else:
        res0 = f(s, i + 1, (d + 1) % 4, _x, _y)
        res1 = f(s, i + 1, (d + 3) % 4, _x, _y)
        return res0 or res1


if (f(s, 0, 0, 0, 0)):
    print('Yes')
else:
    print('No')
